# Started on Tuesday, June 10th, 2019
# by Rodrigo García-López for Adrian Ochoa-Leyva's Metagenomics and Metatranscriptomics Lab at IBt, UNAM, Cuernavaca, Mexico.
# Under GNU GPLv3 license
# This is a log of the commands used to analyze and compare sequences from V3, V4 and V3-V4 regions of the 16S rRNA from bacteria in specific samples. It is not intended to work as a pipeline and will therefore require user input and interpretation. It is provided upon request for the journal Microorganisms. Most commands do not run interactively but print sets of bash commands for easier automated processing. Accessory R script are provided separtely.
# Disclamer: These commands were used for specific data and are not therefore warrantied to be usable with different sets. They were created to carry out particular analyses and were originally intended for the lab's requirements, not commercial nor distribution.
# These commands were used for clustering and analyzing the data

################### QIIME import ###################
# Create a list file with all samples including all those in the V3, V4, and V3-V4 sets.
echo sample-id,absolute-filepath,direction >06_qiime_import/06_Final_mix_manifest.csv; for sample in $(cat Rarefied_16793_06_Final_mix.list);do echo $sample,\$PWD/05.1_New_joined_sets/06_Final_mix/$sample.fastq.gz,forward;done|sed -e 's/_/./'  >>06_qiime_import/06_Final_mix_manifest.csv
echo sample-id,absolute-filepath,direction >06_qiime_import/06_Final_mix_V3_manifest.csv;cat 06_qiime_import/06_Final_mix_manifest.csv|egrep "^V3\." >>06_qiime_import/06_Final_mix_V3_manifest.csv
echo sample-id,absolute-filepath,direction >06_qiime_import/06_Final_mix_V4_manifest.csv;cat 06_qiime_import/06_Final_mix_manifest.csv|egrep "^V4\." >>06_qiime_import/06_Final_mix_V4_manifest.csv
echo sample-id,absolute-filepath,direction >06_qiime_import/06_Final_mix_V3V4_manifest.csv;cat 06_qiime_import/06_Final_mix_manifest.csv|egrep "^V3V4\." >>06_qiime_import/06_Final_mix_V3V4_manifest.csv
# Create a list of all subgroups for automated processing
ls 06_qiime_import/*.csv|cut -d "/" -f 2|sed 's/_manifest.csv//'|grep "V" >subsampled_groups.list
# Initialize the qiime2 python env (with conda). Alternately, use the alias I created: qiime2
source activate qiime2-2019.1
# now create instructions to import the sets
for i in $(cat subsampled_groups.list);do echo date\;qiime tools import --type \'SampleData[SequencesWithQuality]\' --input-path 06_qiime_import/$i\_manifest.csv --output-path 06_qiime_import/$i.qza --input-format SingleEndFastqManifestPhred33\;date;done

################### Create OTUs/ASvs ###################
 ### Dereplicate samples ###
# The joined sequences will be used for this clustering step. The reference to be used in this step will be a 97% clustered SILVA or greengenes.
mkdir -p 07.1_pick_otus_open_ref/01_dereplication
# Run dereplication for each group/set combination
for i in $(cat subsampled_groups.list);do echo date\;qiime vsearch dereplicate-sequences --i-sequences 06_qiime_import/$i.qza --o-dereplicated-table 07.1_pick_otus_open_ref/01_dereplication/$i\_table.qza --o-dereplicated-sequences 07.1_pick_otus_open_ref/01_dereplication/$i\_rep-seqs.qza\;date;done
 ### Carry out open-references clustering with vsearch ###
# There's no need to repeat the preparatio nof the reference databases so I'll just load that from the other project folder: /home/rodrigo/samples_2015-2018/16S/07.1_pick_otus_open_ref/02_16S_Ref_Clusters_and_Training_reference/Greengenes/ and /home/rodrigo/samples_2015-2018/16S/07.1_pick_otus_open_ref/02_16S_Ref_Clusters_and_Training_reference/Silva/
# Carry out open reference clustering
mkdir -p 07.1_pick_otus_open_ref/03_cluster_open_reference
# With greengenes
for i in $(cat subsampled_groups.list);do echo qiime vsearch cluster-features-open-reference --i-table 07.1_pick_otus_open_ref/01_dereplication/$i\_table.qza --i-sequences 07.1_pick_otus_open_ref/01_dereplication/$i\_rep-seqs.qza --i-reference-sequences /home/rodrigo/samples_2015-2018/16S/07.1_pick_otus_open_ref/02_16S_Ref_Clusters_and_Training_reference/Greengenes/97_gg_ref-seqs.qza --p-perc-identity 0.97 --output-dir 07.1_pick_otus_open_ref/03_cluster_open_reference/$i\_gg97 --p-threads 4 --p-strand both;done
# With silva (the DB is larger so we can process the reference-based part using more cpus)
for i in $(cat subsampled_groups.list);do echo qiime vsearch cluster-features-open-reference --i-table 07.1_pick_otus_open_ref/01_dereplication/$i\_table.qza --i-sequences 07.1_pick_otus_open_ref/01_dereplication/$i\_rep-seqs.qza --i-reference-sequences /home/rodrigo/samples_2015-2018/16S/07.1_pick_otus_open_ref/02_16S_Ref_Clusters_and_Training_reference/Silva/97_silva_ref-seqs.qza --p-perc-identity 0.97 --output-dir 07.1_pick_otus_open_ref/03_cluster_open_reference/$i\_silva97 --p-threads 10 --p-strand both;done
### Remove singletons ###
# By default, the open-reference clustering does not remove singletons, so we manually remove them.
# I changed the name of the folder to avoid changing all the downstream commands
mv 07.1_pick_otus_open_ref/03_cluster_open_reference 07.1_pick_otus_open_ref/with_singletons_03_cluster_open_reference
# Now filter all tables to remove singletons and create new rep-seq sets by filtering the clusters with the newly generated tables
for i in $(cat subsampled_groups.list);do for j in gg97 silva97; do echo 'mkdir -p 07.1_pick_otus_open_ref/03_cluster_open_reference/'$i'_'$j'; qiime feature-table filter-features --i-table 07.1_pick_otus_open_ref/with_singletons_03_cluster_open_reference/'$i'_'$j'/clustered_table.qza --p-min-frequency 2 --o-filtered-table 07.1_pick_otus_open_ref/03_cluster_open_reference/'$i'_'$j'/clustered_table.qza; qiime feature-table filter-seqs --i-data 07.1_pick_otus_open_ref/with_singletons_03_cluster_open_reference/'$i'_'$j'/clustered_sequences.qza --i-table 07.1_pick_otus_open_ref/03_cluster_open_reference/'$i'_'$j'/clustered_table.qza --o-filtered-data 07.1_pick_otus_open_ref/03_cluster_open_reference/'$i'_'$j'/clustered_sequences.qza';done;done
 ### Filter reference-based chimeric sequences ###
# We used the singleton-free-clusters to get reference-based chimeras
# For this, we'll be using the gold.fa reference from the broad institute at /home/rodrigo/samples_2015-2018/16S/07.1_pick_otus_open_ref/02_16S_Ref_Clusters_and_Training_reference/Gold.qza
mkdir -p 07.1_pick_otus_open_ref/04_uchime_chimera_removal/01_ref
for i in $(cat subsampled_groups.list);do for j in gg97 silva97; do echo 'time qiime vsearch uchime-ref --i-table 07.1_pick_otus_open_ref/03_cluster_open_reference/'$i'_'$j'/clustered_table.qza --i-sequences 07.1_pick_otus_open_ref/03_cluster_open_reference/'$i'_'$j'/clustered_sequences.qza --i-reference-sequences /home/rodrigo/samples_2015-2018/16S/07.1_pick_otus_open_ref/02_16S_Ref_Clusters_and_Training_reference/Gold.qza --output-dir 07.1_pick_otus_open_ref/04_uchime_chimera_removal/01_ref/'$i'_'$j' --p-minh .3 --p-threads 10';done;done
mkdir -p 07.1_pick_otus_open_ref/04_uchime_chimera_removal/02_denovo
# We used the singleton-free-clusters to get de novo chimeras
for i in $(cat subsampled_groups.list);do for j in gg97 silva97; do echo 'time qiime vsearch uchime-denovo --i-table 07.1_pick_otus_open_ref/03_cluster_open_reference/'$i'_'$j'/clustered_table.qza --i-sequences 07.1_pick_otus_open_ref/03_cluster_open_reference/'$i'_'$j'/clustered_sequences.qza --output-dir 07.1_pick_otus_open_ref/04_uchime_chimera_removal/02_denovo/'$i'_'$j' --p-minh .3';done;done
 ### Retain overlap of ref-based chimeras and de novo chimeras ###
# Now we compare the ref-based chimeras and de novo chimeras to keep just the intersection of both methods
mkdir -p 07.1_pick_otus_open_ref/04_uchime_chimera_removal/03_both_in_denovo_and_ref/
for i in $(cat subsampled_groups.list);do for j in gg97 silva97; do echo 'time qiime feature-table filter-seqs --i-data 07.1_pick_otus_open_ref/04_uchime_chimera_removal/02_denovo/'$i'_'$j'/chimeras.qza --m-metadata-file 07.1_pick_otus_open_ref/04_uchime_chimera_removal/01_ref/'$i'_'$j'/chimeras.qza --o-filtered-data 07.1_pick_otus_open_ref/04_uchime_chimera_removal/03_both_in_denovo_and_ref/'$i'_'$j'-intersect_chimeras.qza';done;done
 ### Filter chimeras from clusters ###
mkdir -p 07.1_pick_otus_open_ref/04_uchime_chimera_removal/04_chimera_free
# We filter the cluster table, the rep-set collection and create a summarized visualization
for i in $(cat subsampled_groups.list);do for j in gg97 silva97; do echo 'qiime feature-table filter-features --i-table 07.1_pick_otus_open_ref/03_cluster_open_reference/'$i'_'$j'/clustered_table.qza --m-metadata-file 07.1_pick_otus_open_ref/04_uchime_chimera_removal/03_both_in_denovo_and_ref/'$i'_'$j'-intersect_chimeras.qza --p-exclude-ids --o-filtered-table 07.1_pick_otus_open_ref/04_uchime_chimera_removal/04_chimera_free/'$i'_'$j'-table-nonchimeric.qza; qiime feature-table filter-seqs --i-data 07.1_pick_otus_open_ref/03_cluster_open_reference/'$i'_'$j'/clustered_sequences.qza --m-metadata-file 07.1_pick_otus_open_ref/04_uchime_chimera_removal/03_both_in_denovo_and_ref/'$i'_'$j'-intersect_chimeras.qza --p-exclude-ids --o-filtered-data 07.1_pick_otus_open_ref/04_uchime_chimera_removal/04_chimera_free/'$i'_'$j'-rep-seqs-nonchimeric.qza; qiime feature-table summarize --i-table 07.1_pick_otus_open_ref/04_uchime_chimera_removal/04_chimera_free/'$i'_'$j'-table-nonchimeric.qza --o-visualization 07.1_pick_otus_open_ref/04_uchime_chimera_removal/04_chimera_free/'$i'_'$j'-table-nonchimeric.qzv';done;done

 ################ Merge sets and compare overlapping OTUs ####################
# Extract the tables and append a set identifier to the OTU_ID and import them with qiime2 format
mkdir -p 07.1_pick_otus_open_ref/04_uchime_chimera_removal/05_tables_merged_per_group/01_tsv_tables
for group in 06_Final_mix;do for region in V3 V4 V3V4;do for ref in gg97 silva97; do out_dir="07.1_pick_otus_open_ref/04_uchime_chimera_removal/05_tables_merged_per_group/01_tsv_tables"; echo 'qiime tools export --input-path 07.1_pick_otus_open_ref/04_uchime_chimera_removal/04_chimera_free/'$group'_'$region'_'$ref'-table-nonchimeric.qza --output-path '$out_dir'; biom convert -i '$out_dir'/feature-table.biom -o '$out_dir'/temp.tsv --to-tsv; perl -ne '\''$i++; print "$_" if ($i==2);print "'$region'.$_" if($i>2)'\'' '$out_dir'/temp.tsv >'$out_dir'/'$group'_'$region'_'$ref'_table-nonchimeric.tsv; biom convert -i '$out_dir'/'$group'_'$region'_'$ref'_table-nonchimeric.tsv -o '$out_dir'/out.biom --table-type="OTU table" --to-hdf5; qiime tools import --input-path '$out_dir'/out.biom --type '\''FeatureTable[Frequency]'\'' --input-format BIOMV210Format --output-path '$out_dir'/'$group'_'$region'_'$ref'_table-nonchimeric';done;done;done
# Now extract the fastas and append a set identifier to the OTU_ID
mkdir -p 07.1_pick_otus_open_ref/04_uchime_chimera_removal/05_tables_merged_per_group/02_fastas
for group in 06_Final_mix;do for ref in gg97 silva97; do for region in V3 V4 V3V4;do out_dir="07.1_pick_otus_open_ref/04_uchime_chimera_removal/05_tables_merged_per_group"; echo 'qiime tools export --input-path 07.1_pick_otus_open_ref/04_uchime_chimera_removal/04_chimera_free/'$group'_'$region'_'$ref'-rep-seqs-nonchimeric.qza --output-path '$out_dir'/02_fastas;  sed "s/^>/>'$region'./" '$out_dir'/02_fastas/dna-sequences.fasta >>'$out_dir'/02_fastas/'$group'_'$region'_'$ref'-rep-seqs-nonchimeric.fasta';done;done;done
# We can now merge the fastas per group/reference set
mkdir -p 07.1_pick_otus_open_ref/04_uchime_chimera_removal/05_tables_merged_per_group/03_joined_no_overlap
for group in 06_Final_mix;do for ref in gg97 silva97; do for region in V3 V4 V3V4;do out_dir="07.1_pick_otus_open_ref/04_uchime_chimera_removal/05_tables_merged_per_group"; echo 'cat '$out_dir'/02_fastas/'$group'_'$region'_'$ref'-rep-seqs-nonchimeric.fasta >>'$out_dir'/03_joined_no_overlap/'$group'_'$ref'-rep-seqs-nonchimeric.fasta';done;done;done
# And import them for qiime2
out_dir="07.1_pick_otus_open_ref/04_uchime_chimera_removal/05_tables_merged_per_group/03_joined_no_overlap"; for i in $(ls $out_dir);do echo 'qiime tools import --input-path '$out_dir'/'$i' --type '\''FeatureData[Sequence]'\'' --output-path '$out_dir'/'${i%?????}'qza';done
# Since we created a qza file for each table, we can now merge them in qiime2 (we could not do this previously because OTU_IDs were equivalent in the V3, V4, and V3V4 sets, now we have unique identifiers and can avoid summing them.
out_dir="07.1_pick_otus_open_ref/04_uchime_chimera_removal/05_tables_merged_per_group/";for group in 06_Final_mix;do for ref in gg97 silva97; do echo 'qiime feature-table merge --i-tables '$out_dir'/01_tsv_tables/'$group'_V3_'$ref'_table-nonchimeric.qza --i-tables '$out_dir'/01_tsv_tables/'$group'_V4_'$ref'_table-nonchimeric.qza --i-tables '$out_dir'/01_tsv_tables/'$group'_V3V4_'$ref'_table-nonchimeric.qza --o-merged-table '$out_dir'/03_joined_no_overlap/'$group'_'$ref'_table-nonchimeric.qza; qiime tools export --input-path '$out_dir'/03_joined_no_overlap/'$group'_'$ref'_table-nonchimeric.qza --output-path '$out_dir'/03_joined_no_overlap/; biom convert -i '$out_dir'/03_joined_no_overlap/feature-table.biom -o '$out_dir'/03_joined_no_overlap/'$group'_'$ref'_table-nonchimeric.tsv --to-tsv';done;done
#This command created non-overlapping tables where sets V3, V4, and V3V4 are processed independently. These are directly usable for downstream processes, including phylogenetic-based metrics.
# The name of the joined folder was changed to 03_joined_no_overlap
# However, it is still useful to compare reference-based clusters that intersect the sets. This can be done by using the separate V3, V4, and V3V4 and joining them with qiime's tools before the names are actually altered. This will not allow for further phylogenetic-based comparisons but will give an insight into which sequences hit the same reference in each set and their simultaneously matching items.
mkdir 07.1_pick_otus_open_ref/04_uchime_chimera_removal/05_tables_merged_per_group/04_joined_with_ref-based_overlap
out_dir="07.1_pick_otus_open_ref/04_uchime_chimera_removal"; for group in 06_Final_mix;do for ref in gg97 silva97; do echo 'qiime feature-table merge --i-tables '$out_dir'/04_chimera_free/'$group'_V3_'$ref'-table-nonchimeric.qza --i-tables '$out_dir'/04_chimera_free/'$group'_V4_'$ref'-table-nonchimeric.qza --i-tables '$out_dir'/04_chimera_free/'$group'_V3V4_'$ref'-table-nonchimeric.qza --o-merged-table '$out_dir'/05_tables_merged_per_group/04_joined_with_ref-based_overlap/'$group'_'$ref'_table-nonchimeric.qza; qiime tools export --input-path '$out_dir'/05_tables_merged_per_group/04_joined_with_ref-based_overlap/'$group'_'$ref'_table-nonchimeric.qza --output-path '$out_dir'/05_tables_merged_per_group/04_joined_with_ref-based_overlap/; biom convert -i '$out_dir'/05_tables_merged_per_group/04_joined_with_ref-based_overlap/feature-table.biom -o '$out_dir'/05_tables_merged_per_group/04_joined_with_ref-based_overlap/'$group'_'$ref'_table-nonchimeric.tsv --to-tsv';done;done

################### Taxonomic classification ##################
# We'll be using the training models for greengenes 97 and 99% clusters V3-V4 regions with sklearn created for other project, found at: /home/rodrigo/samples_2015-2018/16S/07.1_pick_otus_open_ref/02_16S_Ref_Clusters_and_Training_reference/Greengenes
# NOTE: Representative sequences merged without previous cluster name relabeling (04_joined_with_ref-based_overlap) cannot be used here. They have no accurate way to assign a single representative sequence when there is a simultaneous match accross the V3, V4 and/or V3V4.
 ### OTU classification ###
mkdir -p 08_taxonomic_classification/OTUs_97
# Greengenes
in_dir="07.1_pick_otus_open_ref/04_uchime_chimera_removal/05_tables_merged_per_group/03_joined_no_overlap";for group in 06_Final_mix;do echo 'qiime feature-classifier classify-sklearn --i-classifier /home/rodrigo/samples_2015-2018/16S/07.1_pick_otus_open_ref/02_16S_Ref_Clusters_and_Training_reference/Greengenes/97_gg_classifier.qza --i-reads '$in_dir'/'$group'_gg97-rep-seqs-nonchimeric.qza --o-classification 08_taxonomic_classification/OTUs_97/'$group'_gg97-taxonomy.qza --p-n-jobs 20';done
# Silva
in_dir="07.1_pick_otus_open_ref/04_uchime_chimera_removal/05_tables_merged_per_group/03_joined_no_overlap/";for group in 06_Final_mix;do echo 'qiime feature-classifier classify-sklearn --i-classifier /home/rodrigo/samples_2015-2018/16S/07.1_pick_otus_open_ref/02_16S_Ref_Clusters_and_Training_reference/Silva/97_silva_classifier.qza --i-reads '$in_dir'/'$group'_silva97-rep-seqs-nonchimeric.qza --o-classification 08_taxonomic_classification/OTUs_97/'$group'_silva97-taxonomy.qza --p-n-jobs 60';done
 ### Get OTUs-level tables ###
# Get contingency tables with taxonomy for the raw OTU counts (qiime2 v2019.1 does not provide this feature yet)
# extract the raw OTU tables without taxonomy (the merged tables from the rep_set minus the chimeric cluster seqs). This has the abundance info in a binary table.
# NOTE 1: Please note we'll be working with tables that were merged here. For these, everything, from clusters to chimera filterings were processsed separately by region (V3, V4, and V3-V4). Thus, we will have several repeated taxonomies accross clusters. Tables and taxonomy were relabeled in previous steps in order to trace the region (this have no overlaps between them). 
# NOTE 2: Tables merged without previous cluster name relabeling do not have an assigned taxonomy, as several items occur simultaneously accross the region sets but no single sequence can be used for determining taxonomy.
# The following steps are only required for the OTU lvl tables
# We need to extract the taxonomy (tsv) and the table files (biom) and fix their headers in the former, then append the corresponding taxonomy to the biom tables and output the tables as tsv
in_dir="07.1_pick_otus_open_ref/04_uchime_chimera_removal/05_tables_merged_per_group/03_joined_no_overlap";out_dir="09_Raw_contingency_tables/OTUs"; for group in 06_Final_mix;do for ref in gg silva; do mkdir -p 09_Raw_contingency_tables/OTUs/$ref; echo 'qiime tools export --input-path '$in_dir'/'$group'_'$ref'97_table-nonchimeric.qza --output-path temp_tables; qiime tools export --input-path 08_taxonomic_classification/OTUs_97/'$group'_'$ref'97-taxonomy.qza --output-path temp_tables;cat <(printf "#OTUID\ttaxonomy\tconfidence\n") <(tail -n +2 temp_tables/taxonomy.tsv) >temp_tables/temp;mv temp_tables/temp temp_tables/taxonomy.tsv; biom add-metadata -i temp_tables/feature-table.biom -o '$out_dir'/'$ref'/'$group'-table_lvlOTU.biom --observation-metadata-fp temp_tables/taxonomy.tsv --sc-separated taxonomy; biom convert -i '$out_dir'/'$ref'/'$group'-table_lvlOTU.biom -o '$out_dir'/'$ref'/'$group'_'$ref'97-table_lvlOTU_w_tax.tsv --to-tsv --header-key taxonomy';done;done
# Create an additional OTU set without the taxonomy
out_dir="09_Raw_contingency_tables/OTUs"; for group in 06_Final_mix;do for ref in gg silva; do echo 'cat '$out_dir'/'$ref'/'$group'_'$ref'97-table_lvlOTU_w_tax.tsv|rev|cut -f 2-|rev >'$out_dir'/'$ref'/'$group'_'$ref'97-table_lvlOTU.tsv';done;done
 ### Summarize tables ###
# We can now get 7 new tables summarized by taxonomy
#OTUs
in_dir="07.1_pick_otus_open_ref/04_uchime_chimera_removal/05_tables_merged_per_group/03_joined_no_overlap";out_dir="09_Raw_contingency_tables/OTUs"; for ref in gg silva; do for group in 06_Final_mix; do for lvl in {1..7}; do echo 'qiime taxa collapse --i-table '$in_dir'/'$group'_'$ref'97_table-nonchimeric.qza --i-taxonomy 08_taxonomic_classification/OTUs_97/'$group'_'$ref'97-taxonomy.qza --p-level '$lvl' --o-collapsed-table '$out_dir'/'$ref'/'$group'_'$ref'97-table_lvl'$lvl'; qiime tools export --input-path '$out_dir'/'$ref'/'$group'_'$ref'97-table_lvl'$lvl'.qza --output-path '$out_dir'/'$ref'/'$group'_'$ref'97-biom'$lvl'; biom convert -i '$out_dir'/'$ref'/'$group'_'$ref'97-biom'$lvl'/feature-table.biom -o '$out_dir'/'$ref'/'$group'_'$ref'97-table_lvl'$lvl'.tsv --to-tsv';done;done;done

################### Reduce data sparsity (filter contingency tables) ##################
# Singletons and scarce items artificially increase variability in the sample due to how disperse data are in the table (sparse data). Statistic power relies on comparable items, those single OTUs that are detected in different samples.
# Thus, we want to remove singletons and perhaps a little more, those sequences occuring merely by chance.
# We will be using the separated tables for V3, V4, and V3V4 in each group as representative sequences (centroids) are different and vary differently. These are found in 07.1_pick_otus_open_ref/04_uchime_chimera_removal/05_tables_merged_per_group/01_tsv_tables/. For each table we must get the total items in the whole table (per group/region) to calculate 0.01% (0.0001).
# Now use the calculate_sparsity_parameters.R script to determine the cutoffs and filter the tables using qiime
mkdir 10_Reduced_sparsity_tables/OTUs_97/02.5_Filtered_tables_split_region
out_dir="10_Reduced_sparsity_tables/OTUs_97/02.5_Filtered_tables_split_region"; in_dir="07.1_pick_otus_open_ref/04_uchime_chimera_removal/05_tables_merged_per_group/01_tsv_tables";for group in 06_Final_mix; do for ref in gg silva; do for region in V3 V4 V3V4;do echo 'freq=$(cat '$in_dir/$group'_'$region'_'$ref'97_table-nonchimeric.tsv|~/bin/R-3.6.0/bin/Rscript calculate_sparsity_parameters.R|grep "Freq"|cut -f 2); sample=$(cat '$in_dir/$group'_'$region'_'$ref'97_table-nonchimeric.tsv|~/bin/R-3.6.0/bin/Rscript calculate_sparsity_parameters.R|grep "Sample"|cut -f 2); printf "'$group'\t'$ref'\t'$region'\tMin_Freq:\t$freq\tMin_Samples:\t$sample\n" >>10_Reduced_sparsity_tables/OTUs_97/Sparsity_parameters.tsv;qiime feature-table filter-features --i-table '$in_dir/$group'_'$region'_'$ref'97_table-nonchimeric.qza --p-min-frequency $freq --p-min-samples $sample --o-filtered-table '$out_dir/$group'_'$region'_'$ref'97-table_filtered.qza';done;done;done
# Next, merge the filtered tables (no overlapping sequences should be present as tables already make a distinction of each region belonging to a region)
mkdir 10_Reduced_sparsity_tables/OTUs_97/gg 10_Reduced_sparsity_tables/OTUs_97/silva
in_dir="10_Reduced_sparsity_tables/OTUs_97"; for ref in gg silva; do for group in 06_Final_mix; do echo 'qiime feature-table merge --i-tables '$in_dir'/02.5_Filtered_tables_split_region/'$group'_V3_'$ref'97-table_filtered.qza --i-tables '$in_dir'/02.5_Filtered_tables_split_region/'$group'_V4_'$ref'97-table_filtered.qza --i-tables '$in_dir'/02.5_Filtered_tables_split_region/'$group'_V3V4_'$ref'97-table_filtered.qza --o-merged-table '$in_dir/$ref/$group'_'$ref'97-table_lvlOTU.qza';done;done
 ### OTU table with taxonomy ###
# The following steps are only required for the OTU lvl tables
# We need to extract the taxonomy (tsv) and the table files (biom) and fix their headers in the former, then append the corresponding taxonomy to the biom tables and output the tables as tsv
# Only the one
in_dir="10_Reduced_sparsity_tables/OTUs_97";for ref in gg silva; do for group in 06_Final_mix;do echo 'qiime tools export --input-path '$in_dir/$ref/$group'_'$ref'97-table_lvlOTU.qza --output-path temp_tables; qiime tools export --input-path 08_taxonomic_classification/OTUs_97/'$group'_'$ref'97-taxonomy.qza --output-path temp_tables;cat <(printf "#OTUID\ttaxonomy\tconfidence\n") <(tail -n +2 temp_tables/taxonomy.tsv) >temp_tables/temp;mv temp_tables/temp temp_tables/taxonomy.tsv; biom add-metadata -i temp_tables/feature-table.biom -o '$in_dir/$ref/$group'_'$ref'97-table_lvlOTU.biom --observation-metadata-fp temp_tables/taxonomy.tsv --sc-separated taxonomy; biom convert -i '$in_dir/$ref/$group'_'$ref'97-table_lvlOTU.biom -o '$in_dir/$ref/$group'_'$ref'97-table_lvlOTU_w_tax.tsv --to-tsv --header-key taxonomy';done;done
# Create an additional OTU set without the taxonomy
out_dir="10_Reduced_sparsity_tables/OTUs_97"; for ref in gg silva; do for group in 06_Final_mix;do echo 'cat '$out_dir'/'$ref'/'$group'_'$ref'97-table_lvlOTU_w_tax.tsv|rev|cut -f 2-|rev >'$out_dir'/'$ref'/'$group'_'$ref'97-table_lvlOTU.tsv';done;done
# And create the collated taxonomy at all 7 taxonomic levels
out_dir="10_Reduced_sparsity_tables/OTUs_97"; for ref in gg silva; do for group in 06_Final_mix; do for lvl in {1..7}; do echo 'qiime taxa collapse --i-table '$out_dir/$ref/$group'_'$ref'97-table_lvlOTU.qza --i-taxonomy 08_taxonomic_classification/OTUs_97/'$group'_'$ref'97-taxonomy.qza --p-level '$lvl' --o-collapsed-table '$out_dir'/'$ref'/'$group'_'$ref'97-table_lvl'$lvl'; qiime tools export --input-path '$out_dir'/'$ref'/'$group'_'$ref'97-table_lvl'$lvl'.qza --output-path '$out_dir'/'$ref'/'$group'_'$ref'97-biom'$lvl'; biom convert -i '$out_dir'/'$ref'/'$group'_'$ref'97-biom'$lvl'/feature-table.biom -o '$out_dir'/'$ref'/'$group'_'$ref'97-table_lvl'$lvl'.tsv --to-tsv';done;done;done
# Now we create new rep-seq sets using the filtered tables
mkdir 10_Reduced_sparsity_tables/OTUs_97/02_Filtered_rep-seq_sets
in_dir="07.1_pick_otus_open_ref/04_uchime_chimera_removal/05_tables_merged_per_group/03_joined_no_overlap";out_dir="10_Reduced_sparsity_tables/OTUs_97";for ref in gg silva; do for group in 06_Final_mix; do echo 'qiime feature-table filter-seqs --i-data '$in_dir/$group'_'$ref'97-rep-seqs-nonchimeric.qza --i-table '$out_dir/$ref/$group'_'$ref'97-table_lvlOTU.qza --o-filtered-data '$out_dir/02_Filtered_rep-seq_sets/$group'_'$ref'97-rep-seq.qza';done;done
 ### Taxonomy per level ###
#Create a summary of total items per set and level
mkdir 10_Reduced_sparsity_tables/OTUs_97/03_summary_uniques
out_dir="10_Reduced_sparsity_tables/OTUs_97";for ref in gg silva; do for group in 06_Final_mix; do echo 'printf "lvl\titems\n" >'$out_dir'/03_summary_uniques/10_Summary_uniques_'$group'_'$ref'.tsv'; for lvl in 1 2 3 4 5 6 7 OTU;do echo 'printf "'$lvl'\t" >>'$out_dir'/03_summary_uniques/10_Summary_uniques_'$group'_'$ref'.tsv; lines=`wc -l '$out_dir/$ref/$group'_'$ref'97-table_lvl'$lvl'.tsv|cut -d " " -f1`;expr $lines - 1 >>'$out_dir'/03_summary_uniques/10_Summary_uniques_'$group'_'$ref'.tsv';done;done;done
 ### Venn Diagrams ###
# Create Venn diagrams for items that have an overlap (i had to modify the script to summarize the table for repeated elements after trimming the region identifier)
# First a set of those that are reference-based 
mkdir -p 10_Reduced_sparsity_tables/OTUs_97/07_joined_with_ref-based_overlap/01-only_Ref-based
in_dir="10_Reduced_sparsity_tables/OTUs_97"; out_dir="10_Reduced_sparsity_tables/OTUs_97/07_joined_with_ref-based_overlap/01-only_Ref-based"; for ref in gg silva; do for group in 06_Final_mix;do echo 'cat '$in_dir/$ref/$group'_'$ref'97-table_lvlOTU.tsv|sed "s/onstructed from biom file//"|perl -ne '\''$i++;if($i>2){$_ =~ s/^[V34]*\.//; print "$_";}else{print $_}'\''|grep -v "[a-z]"|~/bin/R-3.6.0/bin/Rscript Create_Venn_Diagrams_from_table.R; mv Venn_unique.pdf '$out_dir/$group'_'$ref'_Venn_unique.pdf;mv Venn_totals.pdf '$out_dir/$group'_'$ref'_Venn_totals.pdf';done;done
# Next, get the comparison with both ref and de novo clusters
mkdir -p 10_Reduced_sparsity_tables/OTUs_97/07_joined_with_ref-based_overlap/02-Ref_and_de_novo
in_dir="10_Reduced_sparsity_tables/OTUs_97"; out_dir="10_Reduced_sparsity_tables/OTUs_97/07_joined_with_ref-based_overlap/02-Ref_and_de_novo"; for ref in gg silva; do for group in 06_Final_mix;do echo 'cat '$in_dir/$ref/$group'_'$ref'97-table_lvlOTU.tsv|sed "s/onstructed from biom file//"|perl -ne '\''$i++;if($i>2){$_ =~ s/^[V34]*\.//; print "$_";}else{print $_}'\''|~/bin/R-3.6.0/bin/Rscript Create_Venn_Diagrams_from_table.R; mv Venn_unique.pdf '$out_dir/$group'_'$ref'_Venn_unique.pdf;mv Venn_totals.pdf '$out_dir/$group'_'$ref'_Venn_totals.pdf';done;done
# And one more for the de novo only clusters
mkdir -p 10_Reduced_sparsity_tables/OTUs_97/07_joined_with_ref-based_overlap/03_only_de_novo
in_dir="10_Reduced_sparsity_tables/OTUs_97"; out_dir="10_Reduced_sparsity_tables/OTUs_97/07_joined_with_ref-based_overlap/03_only_de_novo"; for ref in gg silva; do for group in 06_Final_mix;do echo 'cat '$in_dir/$ref/$group'_'$ref'97-table_lvlOTU.tsv|sed "s/onstructed from biom file//"|perl -ne '\''$i++;if($i>2){$_ =~ s/^[V34]*\.//; print "$_";}else{print $_}'\''|grep "[a-z#]"|~/bin/R-3.6.0/bin/Rscript Create_Venn_Diagrams_from_table.R; mv Venn_unique.pdf '$out_dir/$group'_'$ref'_Venn_unique.pdf;mv Venn_totals.pdf '$out_dir/$group'_'$ref'_Venn_totals.pdf';done;done
 ### Summary ref and de novo clusters (barplots) ###
# We can summarize the total items per sample in the tabes (Totals and unique clusters should be the same numbers regardless of whether the overlapping sequences or the non overlapping sequence are used [same data, but clusters are separated by region])
in_dir="10_Reduced_sparsity_tables/OTUs_97"; out_dir="10_Reduced_sparsity_tables/OTUs_97"; for group in 06_Final_mix;do echo 'paste <(printf "Ref") <(cat '$in_dir/gg/$group'_gg97-table_lvlOTU.tsv|sed "s/onstructed from biom file//"|~/bin/R-3.6.0/bin/Rscript contingency_table-unique_items_per_sample.R|head -n 1) >'$out_dir/$group'_gg_vs_silva-all_items.tsv'; for ref in gg silva; do echo 'paste <(printf "All_'$ref'") <(cat '$in_dir/$ref/$group'_'$ref'97-table_lvlOTU.tsv|sed "s/onstructed from biom file//"|~/bin/R-3.6.0/bin/Rscript contingency_table-unique_items_per_sample.R|sed -n 2p) >>'$out_dir/$group'_gg_vs_silva-all_items.tsv';done;for ref in gg silva; do echo 'paste <(printf "Ref_'$ref'") <(cat '$in_dir/$ref/$group'_'$ref'97-table_lvlOTU.tsv|sed "s/onstructed from biom file//"|grep -v "[a-z]"|~/bin/R-3.6.0/bin/Rscript contingency_table-unique_items_per_sample.R|sed -n 2p) >>'$out_dir/$group'_gg_vs_silva-all_items.tsv';done;for ref in gg silva; do echo 'paste <(printf "novo_'$ref'") <(cat '$in_dir/$ref/$group'_'$ref'97-table_lvlOTU.tsv|sed "s/onstructed from biom file//"|grep "[a-z#]"|~/bin/R-3.6.0/bin/Rscript contingency_table-unique_items_per_sample.R|sed -n 2p) >>'$out_dir/$group'_gg_vs_silva-all_items.tsv';done;done
# Now we can summarize the tables using an R script to plot with barplots
mkdir -p 10_Reduced_sparsity_tables/OTUs_97/09_barplots_ref_and_de_novo_per_sample
out_dir="10_Reduced_sparsity_tables/OTUs_97";for group in 06_Final_mix;do echo 'cat '$out_dir/$group'_gg_vs_silva-all_items.tsv|~/bin/R-3.6.0/bin/Rscript barplots_cluster_compositions.R; mv barplots_totals_gg.pdf '$out_dir'/09_barplots_ref_and_de_novo_per_sample/'$group'_barplots_totals_gg.pdf; mv barplots_totals_silva.pdf '$out_dir'/09_barplots_ref_and_de_novo_per_sample/'$group'_barplots_totals_silva.pdf; mv barplots_unique_gg.pdf '$out_dir'/09_barplots_ref_and_de_novo_per_sample/'$group'_barplots_unique_gg.pdf; mv barplots_unique_silva.pdf '$out_dir'/09_barplots_ref_and_de_novo_per_sample/'$group'_barplots_unique_silva.pdf';done

################### Phylogenetic reconstruction ##################
# Phylogenies were created with by inserting sOTU sequences from different regions into the same backbone tree with SEPP (from qiime2; written in python).
# Reference (phylogeny and matching alignment) was Greengenes.
# Using the post-sparsity-reduction set (tables filtered of clusters having <0.01% of the total counts.
mkdir -p 11_phylogenies/OTUs_97/02_Reduced_sparsity
in_dir="10_Reduced_sparsity_tables/OTUs_97/02_Filtered_rep-seq_sets";out_dir="11_phylogenies/OTUs_97/02_Reduced_sparsity"; for group in 06_Final_mix;do for ref in gg97 silva97; do echo 'qiime fragment-insertion sepp --i-representative-sequences '$in_dir'/'$group'_'$ref'-rep-seq.qza --o-tree '$out_dir/$group'_'$ref'-tree.qza --o-placements '$out_dir/$group'_'$ref'-placements.qza --p-threads 4';done;done

################### Diversity Analyses ##################
 ### OTUs Recruitment plots ###
# An R script was created for the recruitment analyses. The input is the regular biom table transformed to tsv.
mkdir -p 12_diversity_analyses/02_using_sparsity_reduced_tables/01_rarefaction_recruitment
in_dir="10_Reduced_sparsity_tables/OTUs_97";out_dir="12_diversity_analyses/02_using_sparsity_reduced_tables/01_rarefaction_recruitment"; for group in 06_Final_mix;do for ref in gg silva; do for lvl in 1 2 3 4 5 6 7 OTU;do echo 'mkdir '$out_dir/$group'_'$ref'_lvl'$lvl';cat '$in_dir/$ref/$group'_'$ref'97-table_lvl'$lvl'.tsv|~/bin/R-3.6.0/bin/Rscript create_recruitment_plot_from_table.R; mv Sample_sum_per_set-table.tsv '$out_dir/$group'_'$ref'_lvl'$lvl'/'$group'_'$ref'97_lvl'$lvl'-Sample_sum_per_set-table.tsv; mv Sample_sum_per_set.pdf '$out_dir/$group'_'$ref'_lvl'$lvl'/'$group'_'$ref'97_lvl'$lvl'-Sample_sum_per_set.pdf; mv Sample_sum_quantiles.pdf '$out_dir/$group'_'$ref'_lvl'$lvl'/'$group'_'$ref'97_lvl'$lvl'-Sample_sum_quantiles.pdf; mv rarefaction_cutoff_adjustment.pdf '$out_dir/$group'_'$ref'_lvl'$lvl'/'$group'_'$ref'97_lvl'$lvl'-rarefaction_cutoff_adjustment.pdf; mv Recruitment_per_sample-organ.pdf '$out_dir/$group'_'$ref'_lvl'$lvl'/'$group'_'$ref'97_lvl'$lvl'-Recruitment_per_sample-organ.pdf; mv Recruitment_per_sample-region.pdf '$out_dir/$group'_'$ref'_lvl'$lvl'/'$group'_'$ref'97_lvl'$lvl'-Recruitment_per_sample-region.pdf; mv Raw_recruitment-20K.pdf '$out_dir/$group'_'$ref'_lvl'$lvl'/'$group'_'$ref'97_lvl'$lvl'-Raw_recruitment-20K.pdf; mv Raw_recruitment-All.pdf '$out_dir/$group'_'$ref'_lvl'$lvl'/'$group'_'$ref'97_lvl'$lvl'-Raw_recruitment-All.pdf; mv Recruitment_organ-20K.pdf '$out_dir/$group'_'$ref'_lvl'$lvl'/'$group'_'$ref'97_lvl'$lvl'-Recruitment_organ-20K.pdf; mv Recruitment_organ-All.pdf '$out_dir/$group'_'$ref'_lvl'$lvl'/'$group'_'$ref'97_lvl'$lvl'-Recruitment_organ-All.pdf; mv Recruitment_region-20K.pdf '$out_dir/$group'_'$ref'_lvl'$lvl'/'$group'_'$ref'97_lvl'$lvl'-Recruitment_region-20K.pdf; mv Recruitment_region-All.pdf '$out_dir/$group'_'$ref'_lvl'$lvl'/'$group'_'$ref'97_lvl'$lvl'-Recruitment_region-All.pdf;rm Rplots.pdf';done;done;done
 ### Relative abundance bar charts ###
# We will be plotting the features occupying the largest percentage in bar plots.
# THESE plots are merely descriptive as there is an underlying bias due to the total items each sample has after chimera removal. Thus, in order to make all bars comparable, we need further transformations.
mkdir -p 12_diversity_analyses/01_using_raw_tables/02_rel_abundance
in_dir="09_Raw_contingency_tables/OTUs";out_dir="12_diversity_analyses/01_using_raw_tables/02_rel_abundance"; for ref in gg silva; do for group in 06_Final_mix;do mkdir -p $out_dir/$group\_$ref; for lvl in 1 2 3 4 5 6 7 OTU;do echo 'cat '$in_dir/$ref/$group'_'$ref'97-table_lvl'$lvl'.tsv|~/bin/R-3.6.0/bin/Rscript plot_abundance_from_table.R; mv Rel_abs_abundance.pdf '$out_dir/$group\_$ref/$group\_$ref'_lvl_'$lvl'-Rel_abs_abundance.pdf';done;done;done
#Same but with the post sparsity sets
mkdir -p 12_diversity_analyses/02_using_sparsity_reduced_tables/02_rel_abundance
in_dir="10_Reduced_sparsity_tables/OTUs_97";out_dir="12_diversity_analyses/02_using_sparsity_reduced_tables/02_rel_abundance"; for ref in gg silva; do for group in 06_Final_mix;do mkdir -p $out_dir/$group\_$ref; for lvl in 1 2 3 4 5 6 7 OTU;do echo 'cat '$in_dir/$ref/$group'_'$ref'97-table_lvl'$lvl'.tsv|~/bin/R-3.6.0/bin/Rscript plot_abundance_from_table.R; mv Rel_abs_abundance.pdf '$out_dir/$group\_$ref/$group\_$ref'_lvl_'$lvl'-Rel_abs_abundance.pdf';done;done;done
# To make taxonomies comparable we rarefied 10000 times to the smallest sample size, then carrying out the comparisons as before with the mean values. The rarefied table is printed as an output. Also, this creates a collated version which aggregates the samples by groups by region/organ. This uses the median value.
mkdir -p 12_diversity_analyses/02_using_sparsity_reduced_tables/05_taxonomy_top
in_dir="10_Reduced_sparsity_tables/OTUs_97";out_dir="12_diversity_analyses/02_using_sparsity_reduced_tables/05_taxonomy_top"; for ref in gg silva; do for group in 06_Final_mix;do mkdir -p $out_dir/$group\_$ref; for lvl in 1 2 3 4 5 6 7 OTU;do echo 'cat '$in_dir/$ref/$group'_'$ref'97-table_lvl'$lvl'.tsv|~/bin/R-3.6.0/bin/Rscript plot_abundance_gastric_bypass-taxonomy_comparable.R '$group'_'$ref'97-table_lvl'$lvl'; mv '$group'_'$ref'97-table_lvl'$lvl'_* '$out_dir/$group\_$ref'';done;done;done

##################################### Some other figures ##########################################
 ### Venn diagrams ###
# We can create Venn diagrams for each of the taxonomic levels (the R script was modified in order to ignore most labels)
mkdir -p 10_Reduced_sparsity_tables/OTUs_97/11_venn_multilvl
in_dir="10_Reduced_sparsity_tables/OTUs_97";out_dir="10_Reduced_sparsity_tables/OTUs_97/11_venn_multilvl"; for ref in gg silva; do for group in 06_Final_mix;do for lvl in 1 2 3 4 5 6 7 OTU;do echo 'cat '$in_dir/$ref/$group'_'$ref'97-table_lvl'$lvl'.tsv|sed "s/onstructed from biom file//"|perl -ne '\''$i++;if($i>2){$_ =~ s/^[V34]*\.//; print "$_";}else{print $_}'\''|~/bin/R-3.6.0/bin/Rscript Create_Venn_Diagrams_from_table.R; mv Venn_unique.pdf '$out_dir/$group'_'$ref'_Venn_unique_lvl'$lvl'.pdf;mv Venn_totals.pdf '$out_dir/$group'_'$ref'_Venn_totals_lvl'$lvl'.pdf';done;done;done
 ### Richness and diversity ###
# We estimate different parameters
mkdir -p 12_diversity_analyses/02_using_sparsity_reduced_tables/04_alpha_diversity_gg_vs_silva_R
in_dir="10_Reduced_sparsity_tables/OTUs_97";out_dir="12_diversity_analyses/02_using_sparsity_reduced_tables/04_alpha_diversity_gg_vs_silva_R"; for ref in gg silva; do for group in 06_Final_mix;do for lvl in 1 2 3 4 5 6 7 OTU;do echo 'cat '$in_dir/$ref/$group'_'$ref'97-table_lvl'$lvl'.tsv|sed "s/onstructed from biom file//"|perl -ne '\''$i++;if($i>2){$_ =~ s/^[V34]*\.//; print "$_";}else{print $_}'\''|~/bin/R-3.6.0/bin/Rscript calculate_alpha.R; mv alpha_table.tsv '$out_dir/$group'_'$ref'_alpha_table_lvl'$lvl'.tsv';done;done;done
# Now plot the shannon entropy and chao1 richness estimator
in_dir="12_diversity_analyses/02_using_sparsity_reduced_tables/04_alpha_diversity_R"; for group in 06_Final_mix;do for lvl in 1 2 3 4 5 6 7 OTU;do echo '~/bin/R-3.6.0/bin/Rscript box_plots_alpha.R '$in_dir/$group'_gg_alpha_table_lvl'$lvl'.tsv '$in_dir/$group'_silva_alpha_table_lvl'$lvl'.tsv; mv chao1.pdf '$in_dir/$group'_chao1_lvl'$lvl'.pdf;mv shannon.pdf '$in_dir/$group'_shannon_lvl'$lvl'.pdf';done;done
# Now compare by sample, organ and region (will only work for set with 4 items (intended for 06_Final_mix)
mkdir -p 12_diversity_analyses/02_using_sparsity_reduced_tables/04.5_alpha_diversity_sample_and_groups
in_dir="10_Reduced_sparsity_tables/OTUs_97";out_dir="12_diversity_analyses/02_using_sparsity_reduced_tables/04.5_alpha_diversity_sample_and_groups"; for group in 06_Final_mix; do for ref in gg silva; do for lvl in 2 3 4 5 6 7 OTU;do echo 'cat '$in_dir/$ref/$group'_'$ref'97-table_lvl'$lvl'.tsv|~/bin/R-3.6.0/bin/Rscript alpha_comparison.R 10000 '$group'_'$ref'_lvl'$lvl';mv '$group'_'$ref'_lvl'$lvl'* '$out_dir'';done;done;done
 ### Determine the % of reads in reference clusters ###
mkdir -p 12_diversity_analyses/02_using_sparsity_reduced_tables/06_beta_diversity
in_dir="10_Reduced_sparsity_tables/OTUs_97";out_dir="12_diversity_analyses/02_using_sparsity_reduced_tables/06_beta_diversity";for group in 06_Final_mix;do for ref in gg silva; do mkdir $out_dir/$group\_$ref; for lvl in 2 3 4 5 6 7 OTU;do echo 'cat '$in_dir/$ref/$group'_'$ref'97-table_lvl'$lvl'.tsv|sed "s/^[V34]*\.//"|~/bin/R-3.6.0/bin/Rscript beta_diversity.R 10000 '$group'_'$ref'_lvl'$lvl';mv '$group'_'$ref'_lvl'$lvl'* '$out_dir/$group\_$ref'';done;done;done
#additionally, use a second one for the OTU set with only reference-based items (which are the only ones that may overlap
in_dir="10_Reduced_sparsity_tables/OTUs_97";out_dir="12_diversity_analyses/02_using_sparsity_reduced_tables/06_beta_diversity";for group in 06_Final_mix;do for ref in gg silva; do mkdir $out_dir/$group\_$ref; for lvl in OTU;do echo 'cat '$in_dir/$ref/$group'_'$ref'97-table_lvl'$lvl'.tsv|sed "s/onstructed from biom file//"|grep -v "[a-z]"|sed "s/^[V34]*\.//"|~/bin/R-3.6.0/bin/Rscript beta_diversity.R 10000 '$group'_'$ref'_lvl'$lvl'-ref;mv '$group'_'$ref'_lvl'$lvl'-ref* '$out_dir/$group\_$ref'';done;done;done

 ########################## Unifrac distance matrices ##########################
# We cannot compare the OTUs using phylogenetic information since we have up to 3 variations of the same sequences in the representative sequences (one per regions) but we can still separate them into regions to calculate this. We will require a filtered table (region-wise), a tree for the corresponding sequences, and carry out the beta div calculation step (qiime2) for unifrac metrics.
 ### Phylogenetic reconstruction with separate regions
# The rep-seq sets and tables can be split into region-specific tables (with post-sparsity-reduction sets) to generate beta diversity plots to better determine local differences. This will require the phylogenetic reconstruction for each set separately.
# First, extract the sequence data (from qiime2 env), split them and import them back to qiime2
mkdir temp
in_dir="10_Reduced_sparsity_tables/OTUs_97/02_Filtered_rep-seq_sets";for group in 06_Final_mix;do for ref in gg silva; do echo 'qiime tools export --input-path '$in_dir/$group\_$ref'97-rep-seq.qza --output-path temp; grep --no-group-separator -A 1 "^>V3\." temp/dna-sequences.fasta >temp/V3_rep-seqs.fasta; grep --no-group-separator -A 1 "^>V4\." temp/dna-sequences.fasta >temp/V4_rep-seqs.fasta; grep --no-group-separator -A 1 "^>V3V4\." temp/dna-sequences.fasta >temp/V3V4_rep-seqs.fasta;qiime tools import --input-path temp/V3_rep-seqs.fasta --type '\''FeatureData[Sequence]'\'' --output-path '$in_dir/$group\_$ref'97_V3-rep-seq.qza;qiime tools import --input-path temp/V4_rep-seqs.fasta --type '\''FeatureData[Sequence]'\'' --output-path '$in_dir/$group\_$ref'97_V4-rep-seq.qza;qiime tools import --input-path temp/V3V4_rep-seqs.fasta --type '\''FeatureData[Sequence]'\'' --output-path '$in_dir/$group\_$ref'97_V3V4-rep-seq.qza';done;done
# Now create matching tables for each of these rep_seqs
mkdir -p 10_Reduced_sparsity_tables/OTUs_97/12_split_region_tables
in_dir="10_Reduced_sparsity_tables/OTUs_97";out_dir="10_Reduced_sparsity_tables/OTUs_97/12_split_region_tables"; for group in 06_Final_mix;do for ref in gg silva; do for region in V3 V4 V3V4; do echo 'grep "^#\|^'$region'\." '$in_dir/$ref/$group\_$ref'97-table_lvlOTU.tsv >'$out_dir/$group\_$ref'97-table_lvlOTU-'$region'.tsv; biom convert -i '$out_dir/$group\_$ref'97-table_lvlOTU-'$region'.tsv -o  temp_table.biom --table-type="OTU table" --to-hdf5; qiime tools import --input-path temp_table.biom --type '\''FeatureTable[Frequency]'\'' --input-format BIOMV210Format --output-path '$out_dir/$group\_$ref'97-table_lvlOTU-'$region'';done;done;done
# We can now create the phylogenies for each rep sequences
mkdir -p 11_phylogenies/OTUs_97/02_Reduced_sparsity/01_split_sequences
in_dir="10_Reduced_sparsity_tables/OTUs_97/02_Filtered_rep-seq_sets";out_dir="11_phylogenies/OTUs_97/02_Reduced_sparsity/01_split_sequences"; for group in 06_Final_mix;do for ref in gg97 silva97; do for region in V3 V4 V3V4; do echo 'qiime fragment-insertion sepp --i-representative-sequences '$in_dir'/'$group'_'$ref'_'$region'-rep-seq.qza --o-tree '$out_dir/$group'_'$ref'_'$region'-tree.qza --o-placements '$out_dir/$group'_'$ref'_'$region'-placements.qza --p-threads 4';done;done;done
# The region tables can be allowed to contain de novo items,since there is no longer a table-wide comparison across region sets
mkdir -p 12_diversity_analyses/02_using_sparsity_reduced_tables/07.1_transformed_tables/02_split_region
in_dir="10_Reduced_sparsity_tables/OTUs_97/12_split_region_tables";out_dir="12_diversity_analyses/02_using_sparsity_reduced_tables/07.1_transformed_tables/02_split_region"; for group in 06_Final_mix;do for ref in gg silva; do mkdir $out_dir/$group\_$ref ;for region in V3 V4 V3V4; do echo 'cat '$in_dir/$group'_'$ref'97-table_lvlOTU-'$region'.tsv|~/bin/R-3.6.0/bin/Rscript multi_contingencty_table_transformations.R 10000 0.5 '$out_dir/$group\_$ref/$group'_'$ref'97-tableOTU-'$region'';done;done;done
#Now convert to biom and import to qiime2 for unifrac (use qiime2 env)
out_dir="12_diversity_analyses/02_using_sparsity_reduced_tables/07_transformed_tables/02_split_region"; for group in 06_Final_mix; do for ref in gg silva; do for region in V3 V4 V3V4; do echo 'cp '$out_dir/$group\_$ref/$group'_'$ref'97-tableOTU-'$region'-rar-10000-perm.tsv temp.tsv; sed -i "1s/^/#OTU ID/" temp.tsv; biom convert -i temp.tsv -o out.biom --table-type="OTU table" --to-hdf5; qiime tools import --input-path out.biom --type '\''FeatureTable[Frequency]'\'' --input-format BIOMV210Format --output-path '$out_dir/$group\_$ref/$group'_'$ref'97-tableOTU-'$region'-rar-10000-perm_qiime2; rm temp.tsv out.biom';done;done;done
 ### Create unifrac matrices for the split regions ###
mkdir -p 12_diversity_analyses/02_using_sparsity_reduced_tables/08_distance_matrices/02_split_region
in_dir="12_diversity_analyses/02_using_sparsity_reduced_tables/07_transformed_tables/02_split_region"; out_dir="12_diversity_analyses/02_using_sparsity_reduced_tables/08_distance_matrices/02_split_region";for group in 06_Final_mix; do for ref in gg silva; do mkdir $out_dir/$group\_$ref; for region in V3 V4 V3V4; do echo 'qiime diversity beta-phylogenetic --i-table '$in_dir/$group\_$ref/$group'_'$ref'97-tableOTU-'$region'-rar-10000-perm_qiime2.qza --i-phylogeny 11_phylogenies/OTUs_97/02_Reduced_sparsity/01_split_sequences/'$group\_$ref'97_'$region'-tree.qza --p-metric unweighted_unifrac --o-distance-matrix '$out_dir/$group\_$ref/$group\_$ref'97-tableOTU-'$region'_unweighted_unifrac.qza; qiime diversity beta-phylogenetic --i-table '$in_dir/$group\_$ref/$group'_'$ref'97-tableOTU-'$region'-rar-10000-perm_qiime2.qza --i-phylogeny 11_phylogenies/OTUs_97/02_Reduced_sparsity/01_split_sequences/'$group\_$ref'97_'$region'-tree.qza --p-metric weighted_unifrac --o-distance-matrix '$out_dir/$group\_$ref/$group\_$ref'97-tableOTU-'$region'_weighted_unifrac.qza';done;done;done
# Export the matrices
out_dir="12_diversity_analyses/02_using_sparsity_reduced_tables/08_distance_matrices/02_split_region";for group in 06_Final_mix; do for ref in gg silva; do for region in V3 V4 V3V4; do echo 'qiime tools export --input-path '$out_dir/$group\_$ref/$group\_$ref'97-tableOTU-'$region'_weighted_unifrac.qza --output-path . ;mv distance-matrix.tsv '$out_dir/$group\_$ref/$group\_$ref'97-tableOTU-'$region'_weighted_unifrac.tsv; qiime tools export --input-path '$out_dir/$group\_$ref/$group\_$ref'97-tableOTU-'$region'_unweighted_unifrac.qza --output-path . ;mv distance-matrix.tsv '$out_dir/$group\_$ref/$group\_$ref'97-tableOTU-'$region'_unweighted_unifrac.tsv';done;done;done
# Now create the PCoA and NMDS plots
mkdir -p 12_diversity_analyses/02_using_sparsity_reduced_tables/09_Unifrac_split_regions
in_dir="12_diversity_analyses/02_using_sparsity_reduced_tables/08_distance_matrices/02_split_region";out_dir="12_diversity_analyses/02_using_sparsity_reduced_tables/09_Unifrac_split_regions";for group in 06_Final_mix; do for ref in gg silva; do mkdir $out_dir/$group\_$ref; for region in V3 V4 V3V4; do echo 'cat '$in_dir/$group\_$ref/$group\_$ref'97-tableOTU-'$region'_unweighted_unifrac.tsv|~/bin/R-3.6.0/bin/Rscript multi_ordination_methods.R '$group\_$ref\_$region'-u_unifrac; cat '$in_dir/$group\_$ref/$group\_$ref'97-tableOTU-'$region'_weighted_unifrac.tsv|~/bin/R-3.6.0/bin/Rscript multi_ordination_methods.R '$group\_$ref\_$region'-w_unifrac; mv '$group\_$ref\_$region'* '$out_dir/$group\_$ref'';done;done;done
 ### Create matrices for joined regions ###
# We had already carried out the organ, region and organ/region clustering of using NMDS with bray curtis and jaccard but since the ordination method uses ranks rather than actual distances, both bray and jaccard turned out quite similar. We will now carry out PCoA with these measures. First, we need the similarity matrices.
mkdir -p 12_diversity_analyses/02_using_sparsity_reduced_tables/07_transformed_tables/01_Multiregion
in_dir="10_Reduced_sparsity_tables/OTUs_97"; out_dir="12_diversity_analyses/02_using_sparsity_reduced_tables/07_transformed_tables/01_Multiregion";for group in 06_Final_mix; do for ref in gg silva; do for lvl in 1 2 3 4 5 6 7; do mkdir $out_dir/$group\_$ref\_$lvl; echo 'cat '$in_dir/$ref/$group\_$ref'97-table_lvl'$lvl'.tsv|~/bin/R-3.6.0/bin/Rscript multi_contingencty_table_transformations.R 10000 0.5 '$out_dir/$group\_$ref\_$lvl/$group'_'$ref'-lvl'$lvl'';done;done;done
# The OTU table must be filtered to allow overlap between regions
in_dir="10_Reduced_sparsity_tables/OTUs_97"; out_dir="12_diversity_analyses/02_using_sparsity_reduced_tables/07_transformed_tables/01_Multiregion";for group in 06_Final_mix; do for ref in gg silva; do for lvl in OTU; do mkdir $out_dir/$group\_$ref\_$lvl; echo 'cat '$in_dir/$ref/$group\_$ref'97-table_lvl'$lvl'.tsv|sed "s/onstructed from biom file//"|grep -v "[a-z]"|sed "s/^[V34]*\.//"|~/bin/R-3.6.0/bin/Rscript multi_contingencty_table_transformations.R 10000 0.5 '$out_dir/$group\_$ref\_$lvl/$group'_'$ref'-lvl'$lvl'';done;done;done
# Now use the rarefied matrix to calculate dissimilarity matrices (jaccard/bray curtis; other metrics can be calculated with this R script but must be uncommented accordingly) 
mkdir -p 12_diversity_analyses/02_using_sparsity_reduced_tables/08_distance_matrices/01_Multiregion
in_dir="12_diversity_analyses/02_using_sparsity_reduced_tables/07_transformed_tables/01_Multiregion"; out_dir="12_diversity_analyses/02_using_sparsity_reduced_tables/08_distance_matrices/01_Multiregion"; for group in 06_Final_mix; do for ref in gg silva; do for lvl in 2 3 4 5 6 7 OTU; do mkdir $out_dir/$group\_$ref\_$lvl; echo 'cat '$in_dir/$group\_$ref\_$lvl/$group'_'$ref'-lvl'$lvl'-rar-10000-perm.tsv|~/bin/R-3.6.0/bin/Rscript multi_distance_metrics.R '$group'_'$ref'-lvl'$lvl'-rar;mv '$group'_'$ref'-lvl'$lvl'-rar* '$out_dir/$group\_$ref\_$lvl/'';done;done;done
# Now create the PCoA and NMDS plots
mkdir -p 12_diversity_analyses/02_using_sparsity_reduced_tables/10_PCoA_NMDS_plots
in_dir="12_diversity_analyses/02_using_sparsity_reduced_tables/08_distance_matrices/01_Multiregion";out_dir="12_diversity_analyses/02_using_sparsity_reduced_tables/10_PCoA_NMDS_plots";for group in 06_Final_mix; do for ref in gg silva; do for lvl in 2 3 4 5 6 7 OTU; do mkdir $out_dir/$group\_$ref\_$lvl; echo 'cat '$in_dir/$group\_$ref\_$lvl/$group'_'$ref'-lvl'$lvl'-rar-jaccard.tsv|~/bin/R-3.6.0/bin/Rscript multi_ordination_methods.R '$group\_$ref\_$lvl'-Rarefied-Jaccard; cat '$in_dir/$group\_$ref\_$lvl/$group'_'$ref'-lvl'$lvl'-rar-bray.tsv|~/bin/R-3.6.0/bin/Rscript multi_ordination_methods.R '$group\_$ref\_$lvl'-Rarefied-Bray; mv '$group\_$ref\_$lvl'* '$out_dir/$group\_$ref\_$lvl'';done;done;done

########################## Differential abundance LEfSe ##########################
# We will be using the rarefied matrix produced when analysing the beta diversity for lvl 7 for the gg set, found at /home/rodrigo/samples_2015-2018/V3-V4_comparison/12_diversity_analyses/02_using_sparsity_reduced_tables/06_beta_diversity/06_Final_mix_gg/06_Final_mix_gg_lvl7-rarefied_matrix.tsv
# At the rarefaction step, the tables that were written were before the division /total rarefaction num, which means we have over 10000 summed together. This did not carry over to NMDS ordination, only on the tables and the script has been fixed but the tables remain the same (takes quite some time to run each iteration). Even so, the proportions should remain the same regardless of the division and it is mainly a stetic issue at this point, as it does not have an impact. Yet, we will fix that in R:
R --no-save #open R with no save option
df <- read.table("06_Final_mix_gg_lvl7-rarefied_matrix.tsv", sep="\t", quote="",header=T, skip=0, comment.char='',fill=F)
df <- rowsum(df[2:length(df)], group=df[[1]]) #sum duplicate items by feature (row) name (if any)
df <- as.matrix(df[order(rowSums(df),decreasing=T),]) # sort by most abundant
df <- df/10000 # divide between the rarefaction number (10K in this case)
write.table(df, "06_Final_mix_gg_lvl7-rarefied_matrix-fixed.tsv", sep="\t", quote=FALSE, col.names=NA, row.names=TRUE) # and export it as a tsv file
q() #Exit R
# Prepare set for lefse #
# LEfSe input files were created by editting the lvl 7 table to be similar to this example (abs values allowed):
# bodysite	mucosal	non_mucosal	mucosal
# subsite	oral	gut	oral
# id	1023	1023	1672
# Bacteria	0.99999	0.99999	0.999993
# Bacteria|Actinobacteria	0.311037	0.000864363	0.00446132
sed -i '1s/^/Sample/' 06_Final_mix_gg_lvl7-rarefied_matrix-fixed.tsv # Add a header for the sample row
# We have to add the region and organ, then fix the empty taxonomic labels to at least have the correct level identifiers, then change nomenclature to reflect that used by LEfSe
cat <(printf "Region\tV3\tV3\tV3\tV3\tV3\tV3\tV3\tV3\tV4\tV4\tV4\tV4\tV4\tV4\tV4\tV4\tV3V4\tV3V4\tV3V4\tV3V4\tV3V4\tV3V4\tV3V4\tV3V4\nOrgan\tH\tH\tH\tH\tI\tI\tI\tI\tH\tH\tH\tH\tI\tI\tI\tI\tH\tH\tH\tH\tI\tI\tI\tI\n") <(cat 06_Final_mix_gg_lvl7-rarefied_matrix-fixed.tsv|sed -e 's/;__\t/;s__\t/' -e 's/;__;s__/;g__;s__/' -e 's/;__;g__/;f__;g__/' -e 's/;__;f__/;o__;f__/' -e 's/;__;o__/;c__;o__/' -e 's/;__;c__/;p__;c__/' -e 's/;__;p__/;k__;p__/' -e 's/k__//' -e 's/;p__/|/' -e 's/;c__/|/' -e 's/;o__/|/' -e 's/;f__/|/' -e 's/;g__/|/' -e 's/;s__/|/' -e 's/;t__/|/') >Tax7.tsv
cat <(printf "Region\tV3\tV3\tV3\tV3\tV3\tV3\tV3\tV3\tV4\tV4\tV4\tV4\tV4\tV4\tV4\tV4\tV3V4\tV3V4\tV3V4\tV3V4\tV3V4\tV3V4\tV3V4\tV3V4\nOrgan\tH\tH\tH\tH\tI\tI\tI\tI\tH\tH\tH\tH\tI\tI\tI\tI\tH\tH\tH\tH\tI\tI\tI\tI\n") <(cat 06_Final_mix_gg_lvl7-rarefied_matrix-fixed.tsv|sed -e 's/;__\t/;s__\t/' -e 's/;__;s__/;g__;s__/' -e 's/;__;g__/;f__;g__/' -e 's/;__;f__/;o__;f__/' -e 's/;__;o__/;c__;o__/' -e 's/;__;c__/;p__;c__/' -e 's/;__;p__/;k__;p__/' -e 's/;s__\t/;s__undetermined\t/' -e 's/;g__;/;g__undetermined;/' -e 's/;f__;/;f__undetermined;/' -e 's/;o__;/;o__undetermined;/' -e 's/;c__;/;c__undetermined;/' -e 's/;p__;/;p__undetermined;/' -e 's/k__;/k__undetermined;/' -e 's/k__//' -e 's/;p__/|/' -e 's/;c__/|/' -e 's/;o__/|/' -e 's/;f__/|/' -e 's/;g__/|/' -e 's/;s__/|/' -e 's/;t__/|/') >Tax7.tsv
# Downstream steps are carried out in the galaxy server.
